dhall/src/Dhall.hs

```diff
@@ -1453,4 +1523,84 @@ inputRecord (RecordInputType inputTypeRecord) = InputType makeRecordLit recordTy
     recordType = Record $ declared <$> inputTypeRecord
     makeRecordLit x = RecordLit $ (($ x) . embed) <$> inputTypeRecord
 
+{-| The 'UnionInputType' monoid allows you to build
+    an 'InputType' injector for a Dhall record.
+
+    For example, let's take the following Haskell data type:
+
+> data Status = Queued Natural
+>             | Result Text
+>             | Errored Text
+
+    And assume that we have the following Dhall union that we would like to
+    parse as a @Status@:
+
+> < Result = "Finish succesfully"
+> | Queued : Natural
+> | Errored : Text
+> >
+
+    Our injector has type 'InputType' @Status@, but we can't build that out of any
+    smaller injectors, as 'InputType's cannot be combined.
+    However, we can use an 'InputUnionType' to build an 'InputType' for @Status@:
+
+> injectStatus :: InputType Status
+> injectStatus =
+>   inputUnion
+>     ( inputConstructorWith "Queued"  inject (\case Queued n  -> Just n; _ -> Nothing)
+>    <> inputConstructorWith "Result"  inject (\case Result t  -> Just t; _ -> Nothing)
+>    <> inputConstructorWith "Errored" inject (\case Errored e -> Just e; _ -> Nothing)
+>     )
+
+    Or, since we are simply using the `Inject` instance to inject each branch, we could write
+
+> injectStatus :: InputType Status
+> injectStatus =
+>   inputUnion
+>     ( inputConstructor "Queued"  (\case Queued n  -> Just n; _ -> Nothing)
+>    <> inputConstructor "Result"  (\case Result t  -> Just t; _ -> Nothing)
+>    <> inputConstructor "Errored" (\case Errored e -> Just e; _ -> Nothing)
+>     )
+
+    Note that the resulting 'InputType' is __not total__ unless we
+    appropriately aggregate each possibility.  If we forgot to handle
+    a branch, the 'InputType' will throw a runtime exception when used on
+    an unhandled branch.
+-}
+newtype UnionInputType a
+  = UnionInputType (Dhall.Map.Map Text (Expr Src X, a -> Maybe (Expr Src X)))
+
+instance Contravariant UnionInputType where
+    contramap f (UnionInputType u) = UnionInputType $ (fmap . flip (.)) f <$> u
 
+instance Semigroup (UnionInputType a) where
+    UnionInputType x <> UnionInputType y = UnionInputType (x <> y)
+
+instance Monoid (UnionInputType a) where
+    mempty = UnionInputType mempty
+
+inputUnion :: UnionInputType a -> InputType a
+inputUnion (UnionInputType inputTypeUnion) = InputType makeUnionLit (Union declare)
+  where
+    declare = fst <$> inputTypeUnion
+    makeUnionLit x = fromMaybe (errorWithoutStackTrace unmatched)
```

Comment:
Would it be possible to avoid the partial implementation by changing `UnionInputType` to:

```haskell
data UnionInputType a = UnionInputType (Map Text (Expr Src X)) (a -> Maybe (Map Expr Src X))
```

Reply:
This representation makes more sense, since we only consume injector in the asum'd form.  I don't think that this directly helps us avoid partiality.

However, this representation gives a hint about how we could make it non-partial; we might be able to write a variant of `inputUnion` that gives a fallback option that would be used if anything was `Maybe`.

Reply:
17fd6c3 now incorporates this new representation!  It's actually nicer because we get a derivable Contravariant and almost a derivable Monoid instance (waiting for base-4.12, which has Monoid for `:*:`).

Still curious about what you had in mind for resolving the partiality issue, though! :)